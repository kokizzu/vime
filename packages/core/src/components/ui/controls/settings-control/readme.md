# vime-settings-control

A control for toggling the visiblity of the settings menu. This control is not displayed if no 
settings (`vime-settings`) has been provided for the current player.

## Example

```html
<vime-player>
  <!-- ... -->
  <vime-ui>
    <!-- ... -->
    <vime-controls>
      <vime-settings-control></vime-settings-control>
    </vime-controls>
  </vime-ui>
<vime-player>
```

<!-- Auto Generated Below -->


## Properties

| Property           | Attribute           | Description                                     | Type                             | Default            |
| ------------------ | ------------------- | ----------------------------------------------- | -------------------------------- | ------------------ |
| `hideTooltip`      | `hide-tooltip`      | Whether the tooltip should not be displayed.    | `boolean`                        | `false`            |
| `icon`             | `icon`              | The URL to an SVG element or fragment to load.  | `string`                         | `'#vime-settings'` |
| `tooltipDirection` | `tooltip-direction` | The direction in which the tooltip should grow. | `"left" \| "right" \| undefined` | `undefined`        |


## Dependencies

### Used by

 - [vime-default-controls](../default-controls)

### Depends on

- [vime-control](../control)
- [vime-icon](../../icon)
- [vime-tooltip](../../tooltip)

### Graph
```mermaid
graph TD;
  vime-settings-control --> vime-control
  vime-settings-control --> vime-icon
  vime-settings-control --> vime-tooltip
  vime-default-controls --> vime-settings-control
  style vime-settings-control fill:#f9f,stroke:#333,stroke-width:4px
```

----------------------------------------------

*Built with [StencilJS](https://stenciljs.com/)*
