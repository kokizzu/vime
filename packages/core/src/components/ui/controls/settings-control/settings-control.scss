vime-settings-control {
  &.hidden {
    display: none;
  }

  svg {
    transition: transform 0.3s ease;
  }

  &.active {
    svg {
      // Need `!important` to override scaling transformation applied to `svg`.
      transform: rotate(90deg) !important;
    }
  }
}