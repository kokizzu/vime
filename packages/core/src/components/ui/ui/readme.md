# vime-ui

Simple container that holds a collection of user interface components.

The only important role this component really has is, avoiding overlapping custom UI with the 
native iOS media player UI. Therefore, custom UI is only displayed on iOS if the `playsinline` prop 
is `true`, and the player is not in fullscreen mode.

## Example

```html
<vime-player>
  <!-- ... -->

  <vime-ui>
    <vime-poster />
    <!-- ... -->
  </vime-ui>
</vime-player>
```

<!-- Auto Generated Below -->


## Dependencies

### Used by

 - [vime-default-ui](../default-ui)

### Graph
```mermaid
graph TD;
  vime-default-ui --> vime-ui
  style vime-ui fill:#f9f,stroke:#333,stroke-width:4px
```

----------------------------------------------

*Built with [StencilJS](https://stenciljs.com/)*
