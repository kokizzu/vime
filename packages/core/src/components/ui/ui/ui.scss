vime-ui {
  position: relative;
  pointer-events: none;
  z-index: $z-index-ui;
  width: 100%;

  &.hidden {
    display: none;
  }

  &.video {
    position: absolute;
    top: 0;
    left: 0;
    height: 100%;
  }
}