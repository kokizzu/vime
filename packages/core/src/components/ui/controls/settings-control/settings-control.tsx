import {
  h, Host, Component, Prop, Element,
} from '@stencil/core';
import { openPlayerWormhole } from '../../../core/player/PlayerWormhole';
import { PlayerProp, PlayerProps } from '../../../core/player/PlayerProp';
import { TooltipDirection } from '../../tooltip/types';
import { isUndefined } from '../../../../utils/unit';
import { findUIRoot } from '../../ui/utils';

let idCount = 0;

@Component({
  tag: 'vime-settings-control',
  styleUrl: 'settings-control.scss',
})
export class SettingsControl {
  private id!: string;

  @Element() el!: HTMLVimeSettingsControlElement;

  /**
   * The URL to an SVG element or fragment to load.
   */
  @Prop() icon = '#vime-settings';

  /**
   * The direction in which the tooltip should grow.
   */
  @Prop() tooltipDirection: TooltipDirection;

  /**
   * @internal
   */
  @Prop() menu?: string;

  /**
   * @internal
   */
  @Prop() expanded = false;

  /**
   * @internal
   */
  @Prop() i18n: PlayerProps[PlayerProp.I18N] = {};

  componentWillLoad() {
    idCount += 1;
    this.id = `vime-settings-control-${idCount}`;
    const settings = findUIRoot(this).querySelector('vime-settings');
    settings?.setController(this.id, this.el);
  }

  private hasSettings() {
    return !isUndefined(this.menu);
  }

  render() {
    return (
      <Host
        class={{
          hidden: !this.hasSettings(),
          active: this.hasSettings() && this.expanded,
        }}
      >
        <vime-control
          identifier={this.id}
          menu={this.menu}
          hidden={!this.hasSettings()}
          expanded={this.expanded}
          label={this.i18n.settings}
        >
          <vime-icon href={this.icon} />

          <vime-tooltip
            hidden={this.expanded}
            direction={this.tooltipDirection}
          >
            {this.i18n.settings}
          </vime-tooltip>
        </vime-control>
      </Host>
    );
  }
}

openPlayerWormhole(SettingsControl, [
  PlayerProp.I18N,
]);
