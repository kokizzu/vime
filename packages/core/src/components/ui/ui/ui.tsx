import {
  h, Host, Component, Prop,
} from '@stencil/core';
import { withPlayerContext } from '../../core/player/PlayerContext';
import { PlayerProp, PlayerProps } from '../../core/player/PlayerProp';
import { IS_IOS } from '../../../utils/support';

/**
 * @slot - Used to pass in UI components for the player.
 */
@Component({
  tag: 'vime-ui',
  styleUrl: 'ui.scss',
})
export class UI {
  /**
   * @internal
   */
  @Prop() isVideoView: PlayerProps[PlayerProp.isVideoView] = false;

  /**
   * @internal
   */
  @Prop() playsinline: PlayerProps[PlayerProp.playsinline] = false;

  /**
   * @internal
   */
  @Prop() isFullscreenActive: PlayerProps[PlayerProp.isFullscreenActive] = false;

  render() {
    const canShowCustomUI = !IS_IOS
      || !this.isVideoView
      || (this.playsinline && !this.isFullscreenActive);

    return (
      <Host
        class={{
          hidden: !canShowCustomUI,
          video: this.isVideoView,
        }}
      >
        <div>
          {canShowCustomUI && <slot />}
        </div>
      </Host>
    );
  }
}

withPlayerContext(UI, [
  PlayerProp.isVideoView,
  PlayerProp.playsinline,
  PlayerProp.isFullscreenActive,
]);
