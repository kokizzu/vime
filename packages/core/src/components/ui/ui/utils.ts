import { getElement } from '@stencil/core';
import { HTMLStencilElement } from '@stencil/core/internal';

export const findUIRoot = (ref: any) => {
  let ui = getElement(ref);

  while (!(/^VIME-UI$/.test(ui?.nodeName))) {
    ui = ui.parentElement as HTMLStencilElement;
  }

  return ui as HTMLVimeUiElement;
};
