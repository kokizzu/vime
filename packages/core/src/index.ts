import { PlayerProp, PlayerProps } from './components/core/player/PlayerProp';
import { PlayerDispatcher, createPlayerDispatcher } from './components/core/player/PlayerDispatcher';
import { withPlayerContext, usePlayerContext } from './components/core/player/PlayerContext';
import { PlayerEvent } from './components/core/player/PlayerEvent';
import { findRootPlayer } from './components/core/player/utils';

export * from './components';

export {
  PlayerProp,
  PlayerProps,
  PlayerEvent,
  findRootPlayer,
  withPlayerContext,
  usePlayerContext,
  PlayerDispatcher,
  createPlayerDispatcher,
};
