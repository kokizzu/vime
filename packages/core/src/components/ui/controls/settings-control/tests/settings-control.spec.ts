import { SpecPage } from '@stencil/core/testing';
import { newUISpecPage } from '../../../ui/tests';
import { SettingsControl } from '../settings-control';

let page: SpecPage;
let control: HTMLVimeSettingsControlElement;

beforeEach(async () => {
  ({ page } = await newUISpecPage(
    [SettingsControl],
    '<vime-settings-control />',
  ));

  control = page.root!.querySelector('vime-settings-control')!;
});

it('should be structurally sound', () => {
  expect(control).toMatchSnapshot();
});
