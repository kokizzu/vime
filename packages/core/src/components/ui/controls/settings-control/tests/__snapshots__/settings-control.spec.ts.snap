// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should be structurally sound 1`] = `
<vime-settings-control class="hidden">
  <vime-control hidden="" identifier="vime-settings-control-1" label="Settings">
    <vime-icon href="#vime-settings"></vime-icon>
    <vime-tooltip>
      Settings
    </vime-tooltip>
  </vime-control>
</vime-settings-control>
`;
